openage/convert/opus/opusenc.pyx

```diff
@@ -0,0 +1,406 @@
+# Copyright 2016-2017 the openage authors. See copying.md for legal info.
+
+import time
+from libc.string cimport memcpy, memset
+from cpython.mem cimport PyMem_Malloc, PyMem_Free
+from cpython.bytes cimport PyBytes_FromStringAndSize
+
+from ...log import dbg, spam
+
+from . cimport ogg, opus
+
+def encode(inputdata=None):
+    '''
+    Converts the wav file in the bytes object 'inputdata' to an opusfile
+    and returns it as bytes object. If something fails, a number or an
+    error message is returned.
+    '''
+    inopt = read_wav(inputdata)
+    if not isinstance(inopt, dict):
+        return inopt
+
+    dbg("Wavefile: Total length: {}\tHeader length: {}\tTrailer length: {}"\
+            .format(len(inputdata), inopt['header_len'], inopt['trailer_len']))
+    if inopt['trailer_len']:
+        inputdata = inputdata[inopt['header_len']:-inopt['trailer_len']]
+    else:
+        inputdata = inputdata[inopt['header_len']:]
+
+    outdata = bytearray(1024*50)
+    cdef size_t outsize = 0 # valid bytes inside 'outdata'
+
+    cdef ogg.ogg_packet op
+    cdef ogg.ogg_page og
+    cdef ogg.ogg_stream_state os
+    cdef opus.OpusEncoder *oe
+    cdef int err
+    cdef int channels = inopt['channels']
+    cdef int wframe_len = inopt['wframe_len']
+    cdef opus.opus_int32 rate = inopt['input_rate']
+    cdef opus.opus_int32 coding_rate = 48000
+    cdef opus.opus_int32 frame_sz = 960
+    cdef opus.opus_int32 lookahead
+
+    if rate > 24000:
+        coding_rate = 48000
+    elif rate > 16000:
+        coding_rate = 24000
+    elif rate > 12000:
+        coding_rate = 16000
+    elif rate > 8000:
+        coding_rate = 12000
+    else:
+        coding_rate = 8000
+
+    frame_sz = <opus.opus_int32> (frame_sz / (48000 / coding_rate))
+
+    if coding_rate != rate:
+        dbg("Resampling necessary: input rate: {}\tcoding rate: {}".format(
+            rate, coding_rate))
+        inputdata = upsample(inputdata, inopt, coding_rate)
+        if not isinstance(inputdata, (bytes, bytearray)):
+            return inputdata
+        dbg("Resampled data length: {}".format(len(inputdata)))
+
+    # initialize structs
+    if ogg.ogg_stream_init(&os, int(time.time() * 100) % (1<<31)):
+        return "Could not initialize ogg_stream."
+
+    oe = opus.opus_encoder_create(coding_rate, channels, opus.OPUS_APPLICATION_AUDIO, &err)
+    if err != opus.OPUS_OK:
+        ogg.ogg_stream_clear(&os)
+        return "Could not initialize opus encoder."
+
+    err = opus.opus_encoder_ctl(oe, opus.OPUS_GET_LOOKAHEAD(&lookahead))
+    if err != opus.OPUS_OK:
+        free_buffers(&os, oe, NULL, NULL)
+        return 1
+
+    inopt['pre_skip'] = lookahead / (48000 / coding_rate)
+
+    if write_opus_header(&os, &op, inopt):
+        free_buffers(&os, oe, NULL, NULL)
```

Comment:
there is a potential memory leak here and in other places. write_opus_header creates a bytearray. if there is no memory left, that fails and a MemoryError is thrown. this is not catched, and free_buffers will not be called.
since this will quit the conversion (and release all memory), I think, this does not need to be fixed.
But I will fix it, if requested.